README.md
Mandrake Update Robot
------------------------------------------------------------------------------
NOTE: This tool is outdated and is no longer developed. Valid from 2000 - 2002, 
deprecated by urpmi. Uploaded to GitHub on August 20, 2016.

License: Freeware (GNU Public License v2) - read file "COPYING"
Author: Prana

CREDITS:
- Patches from Pixel <pixel@mandrakesoft.com>, Geoffrey Lee <snailtalk@linux-mandrake.com>
- Beta testers for corporate proxy/firewall: Kim Schulz <kim@schulz.dk>
- curl-lib (libcurl) by Daniel Stenberg <daniel@haxx.se> - good library!

Mandrake Update Robot is a text-based version of Mandrake Update.
It is very useful if you run a very large network. With this tool, you 
don't have to be worried about missing security updates. After running
the update tool, this robot will e-mail a report to the root user.

See the file "ChangeLog" for release details :-)

Installation:
1) Login as root user
2) type "make"
3) type "make install"
4) run "drakupdaterobot"

Setup:
Run the program "drakupdatesetup" as root user.

Requirements:
* Mandrake Linux 7.2 or higher
* T3/T1/Cable/DSL connection recommended (this drakupdaterobot is not for modem users)
* curl-devel (http://curl.sourceforge.net)


ChangeLog
March 6, 2002
- Fixed - segmentation fault at Stage 5 due to headerFree(rpm_header)
- Fixed - hanging at stage 4

Day 13 - March 31, 2001
- Fixed the curl-lib code in download_file by Daniel Stenberg (author of curl-lib)
- Added more information about custom kernel recompilation
- Fixed the slowness during drakupdatesetup when downloading mirror

Day 12 - March 24, 2001
- Stable for Mandrake 8.x and Mandrake 7.x
- Final version
- Final feature: automatic kernel recompilation
- Options - deleting files after RPM update
- Final bug fixes
- 3069 lines of code

Day 11 - March 21, 2001
- Almost final version 0.9
- New features and bugfixes from 0.8 BETA 3
- Custom FTP mirror selection
- Very fool-proof setup wizard
- More error checking in the network connection during setup wizard
- Change of configuration file format - you have to run drakupdatesetup again
- A more detailed instruction on setting up custom cache/mirror
- Spelling/grammar error has been partially fixed

Day 10 - March 10, 2001
- BETA 3 - focus on quality control
- Started to code custom-build kernel auto-recompilation
- Fixed Squid server support - now it works with authentication ncsa_auth
- Fixed the previous correct download method which was not correct in BETA 2
- Fixed DrakConf entry for Linux Mandrake 7.2
- 1867 lines of code total

Day 9 - March 1, 2001
- Backward compatibility with Mandrake 7.1, 7.2, 1.01 Corporate Server improved
- Compatibility with Mandrake 8.0 is improved
- Compatibility with RPM 4 is improved
- Fix compatibility with Squid proxy server
- Fix major bug - no longer download a package if it's already updated
- Fix weird proxy login password
- Integration with Mandrake 7.2 DrakConf Setup and icon

Day 8 - Feb 24, 2001
- This version is compatible with Linux Mandrake Cooker i586, alpha, sparc,
  and it's backward compatible with Linux Mandrake 7.1 - 7.2, and it's
  compatible with Corporate Server 1.0.1, Linux Mandrake 7.1 on
  Intel i586, Alpha, Sparc SPARC 7.1.
- Designed for the upcoming Linux Mandrake 8.0.
- IMPORTANT: Please delete the older /etc/drakupdatetxt.conf
  from prior versions first before installing this program
- Fixed: MD5 and GnuPG signature checking
- Fixed: Installation and its dependencies
- Fixed: Use HTTP proxy tunnelling for certain proxy servers
- The e-mail report generated by the robot is much more complete,
  with description of updates and vice versa
- DrakUpdateSetup now offers HTTP tunneling option
- Compatibility with RPM 4.0 is improved
- Since I forgot to add rpmTransFree last time, now I added it
- Dependencies added with: mailx, gnupg, crond, anacron, curl-lib
- Added question if the user wants to put it in /etc/cron.daily
- 1765 lines of code total (reduce some stuff, add some stuff)
- Now study for my midterm exams ....

Day 7 - Feb 12, 2001
- Compatibility with the new RPM 4.0 for Linux Mandrake Cooker, sync with DindinX's rpmdrake
- TODO: rpmCheckSig ...

Day 6 - Feb 9, 2001
- Now add: if the RPM is already downloaded but somehow hasn't been installed, install it immediately
- Add: Setup wizard program
- Add: Reading configuration file to customize locations of files
- Add: Support for connection through a proxy server
- Checking for package corruption
- Now approximately 1540 lines of code total...

Day 5 - Feb 5, 2001
- Skipping http mirrors - it will be supported later

Day 4 - Feb 4, 2001
- It's fully functional now: downloading and installing RPMS
- Add RPM installation drakupdate::install_all_selected_rpms()
- Add install_one_rpm_package(string *rpm_package_name) for recursive
- Use recursive function to handle dependencies
- Plan to switch to RPM Library API download function rather than curl-devel for compatibility
- Bug fixes
- Approximately total of 1045 lines of code (not including the removed one) - in 3 hours :)

Day 3 - Jan 31, 2001
- This version day3 is a working prototype of drakupdaterobotz
- Fully Functional: checking which packages needs to be downloaded
- Files that needs to be downloaded will be displayed at the end
- Fixed memory leaks - guaranteed no more memory leaks at this point
- TODO:
  1) actual install process
  2) automatic dependencies after an RPM is downloaded
  3) special kernel installation
  4) e-mail root user for report (replace the 'cerr' with 'ofstream' piece of cake)
  5) Release it ASAP
- Approximately 810 lines of code including comments

Day 2 - Jan 29, 2001
- Use Pixel's suggestion for vector<string> to prevent memory leak
- Fixed huge memory leaks
- Split to 5 different files to make code more maintainable:
 * drakupdaterobot.cpp
 * engine.cpp
 * engine.h
 * util.cpp
 * util.h
- Updated Makefile
- Use MandrakeUpdate's FindMatches and VersionCompare for RPM
- Approximately 850 lines of code

Day 1 - Jan 27, 2001
- Initial development
- Using C++ with Standard Template Library
- Approximately 400 lines of code


util.cpp
// Utilities - util.cpp
/*****************************************************************************
/*  Mandrake Update Text                                                     *
/*  Copyright (C) 2001 Prana                              *
/*                                                                           *
/*                                                                           *
/*  This program is free software; you can redistribute it and/or modify     *
/*  it under the terms of the GNU General Public License as published by     *
/*  the Free Software Foundation; either version 2 of the License, or        *
/*  (at your option) any later version.                                      *
/*                                                                           *
/*  This program is distributed in the hope that it will be useful,          *
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of           *
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the            *
/*  GNU General Public License for more details.                             *
/*                                                                           *
/*  You should have received a copy of the GNU General Public License        *
/*  along with this program; if not, write to the Free Software              *
/*  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111 USA      *
/*****************************************************************************/

#include <stl.h>
#include <curl/curl.h>
#include <curl/types.h>
#include <curl/easy.h>
#include <fstream>
#include <regex.h>
#include <cstdlib>
#include <fcntl.h>

#include <rpm/rpmlib.h>
#include <rpm/rpmio.h>
#include <rpm/misc.h>
#ifndef RPMDBI_PACKAGES
#include <rpm/dbindex.h>
#endif

#include "util.h"

using namespace std;

/*************************************************************************
*  VersionCompare - modified from Mandrake Update
*  March 25, 2001 - added modification for workaround on Redhat's rpmvercmp bug
*                   and Mandrake's packaging bug
**************************************************************************/
int VersionCompare(const char *rpm_new_version, const char *rpm_old_version)
{
  int x, y;
  int return_value;
  // version and release
  string new_ver, old_ver, new_rel, old_rel;
  string NewVersion = rpm_new_version;
  string OldVersion = rpm_old_version;
  
  x = NewVersion.rfind('-');
  y = OldVersion.rfind('-');
  
  new_ver = NewVersion.substr(0,x);
  new_rel = NewVersion.substr(x+1,NewVersion.length());

  old_ver = OldVersion.substr(0,y);
  old_rel = OldVersion.substr(y+1,OldVersion.length());
  
  return_value = rpmvercmp(new_ver.c_str(), old_ver.c_str());
    
  if (return_value == 0)
  {
     return_value = rpmvercmp(new_rel.c_str(), old_rel.c_str());
     if (return_value == -1)
     {
        if ( (NewVersion.find(".1mdk") != -1) && (OldVersion.find(".1mdk")== -1) )
           return_value = 1;
     }
  }
//  cout << "Comparing " << rpm_new_version << " vs" << rpm_old_version << " = " << return_value << endl;  
  return return_value;
}

/****************************************************************************/
bool getversion_and_machine(string &version, string &machine)
// Get version from /etc/mandrake-release
{
   const int SIZE_LIMIT = 512;
   ifstream mandrake_release;
   bool machine_found = false;
   bool return_value = false;
   bool cooker = false;   
   int position = 0;
   int temp = 0;
   char *buffer = new char [SIZE_LIMIT+1];
   // March 21, 2001 - workaround with Beta 8.0
   string tempstr;
   // Read string s from file
   mandrake_release.open(version_file);
   if (mandrake_release)
   {
      for (position = 0; position < 4; position++)
      {
         mandrake_release >> version;
         if (version.find("Cooker") != -1)
         {
            cooker = true;
            temp = version.find("-");
            machine = version.substr(temp+1, version.length());
            mandrake_release >> version;
            break;
         }
      } // end for
      if (!cooker)
      {
         for (position = 0; position < 3; position++)
            mandrake_release >> machine;
      } // end if NOT cooker
      
      // END READING FILE
      mandrake_release.close();
      // March 21, 2001 - Work around with 8.0 bugs ...
      mandrake_release.open(version_file);
      mandrake_release.getline(buffer, SIZE_LIMIT);
      tempstr = buffer;
      mandrake_release.close();

      if (tempstr.find("Cooker") != -1)
      {
         cooker = true; // Can't assign value directly with cooker = mandrake_release.find...
         version = "BETA";
      }
      if (tempstr.find ("i586") != -1) machine = "i586";
      else if (tempstr.find ("alpha") != -1) machine = "alpha";
      // why do you put sparc into cookerparc? That's weird.
      if (machine == "sparc" || (tempstr.find("sparc") != -1))
         machine = "parc";         
      // If it's cooker
      if (cooker)
         machine = "cooker" + machine;
      else
         machine = "updates" + machine;
      return_value = (machine.length() > 0) && (version.length() > 0);
   }
   delete []buffer;   
   return return_value;
}


/****************************************************************************/
int find_in_vector_exact(const vector<string> &the_vector, string &search_for)
{
   bool found = false;
   int return_value = -1;
   int i;
   
   for (i = 0; i < the_vector.size() && !found; i++)
   {
      found = (search_for == (the_vector)[i]);
   }
   if (found)
      return_value = i-1;
   return return_value;
}

/****************************************************************************/
int find_in_vector_name_only(const vector<string> &the_vector, string &search_for)
// e.g: gnome-telnet-2.4 and gnome-telnet-2.3 will return found=true because
// the package name excluding version is the same
{
   bool found = false;
   int return_value = -1;
   int i;
   string tempstr1, tempstr2;
   
   tempstr1 = search_for.substr(0, get_package_name_position(search_for));
   for (i = 0; !found && i < the_vector.size(); i++)
   {
      tempstr2 = (the_vector)[i];
      tempstr2 = tempstr2.substr(0, get_package_name_position(tempstr2));
      found = (tempstr1 == tempstr2);
   }
   if (found)
      return_value = i-1;
   return return_value;
}

int get_package_name_position(const string &packagename)
// Just find the version starting from ...
// for example: "kernel-2.2.17" -> this function will
// return the string position of the first digit found in that text
// Keyword: FIX TODO REPAIR
{
   int i = 0;
   while (i < packagename.length() && !isdigit( packagename[i] ) )
   {
      i++;
   }
   if (i != packagename.length())
      i--;
   return i;
}

void print_curl_error(int error_code)
{
   const char *error_print_format[] = 
   {
     "Error Curl Library: UNSUPPORTED PROTOCOL",
     "Error Curl Library: FAILED INIT",
     "Error Curl Library: URL MALFORMAT",
     "Error Curl Library: URL MALFORMAT USER",
     "Error Curl Library: COULDNT RESOLVE PROXY",
     "Error Curl Library: COULDNT RESOLVE HOST",
     "Error Curl Library: COULDNT CONNECT",
     "Error Curl Library: FTP WEIRD SERVER REPLY",
     "Error Curl Library: FTP ACCESS DENIED",
     "Error Curl Library: FTP USER PASSWORD INCORRECT",
     "Error Curl Library: FTP WEIRD PASS REPLY",
     "Error Curl Library: FTP WEIRD USER REPLY",
     "Error Curl Library: FTP WEIRD PASV REPLY",
     "Error Curl Library: FTP WEIRD 227 FORMAT",
     "Error Curl Library: FTP CANT GET HOST",
     "Error Curl Library: FTP CANT RECONNECT",
     "Error Curl Library: FTP COULDNT SET BINARY",
     "Error Curl Library: PARTIAL FILE",
     "Error Curl Library: FTP COULDNT RETR FILE",
     "Error Curl Library: FTP WRITE ERROR",
     "Error Curl Library: FTP QUOTE ERROR",
     "Error Curl Library: HTTP NOT FOUND",
     "Error Curl Library: WRITE ERROR",

     "Error Curl Library: MALFORMAT USER", /* the user name is illegally specified */
     "Error Curl Library: FTP COULDNT STOR FILE", /* failed FTP upload */
     "Error Curl Library: READ ERROR", /* could open/read from file */

     "Error Curl Library: OUT OF MEMORY",
     "Error Curl Library: OPERATION TIMEOUTED", /* the timeout time was reached */
     "Error Curl Library: FTP COULDNT SET ASCII", /* TYPE A failed */

     "Error Curl Library: FTP PORT FAILED", /* FTP PORT operation failed */

     "Error Curl Library: FTP COULDNT USE REST", /* the REST command failed */
     "Error Curl Library: FTP COULDNT GET SIZE", /* the SIZE command failed */

     "Error Curl Library: HTTP RANGE ERROR", /* The RANGE "command" didn't seem to work */

     "Error Curl Library: HTTP POST ERROR",

     "Error Curl Library: SSL CONNECT ERROR", /* something was wrong when connecting with SSL */

     "Error Curl Library: FTP BAD DOWNLOAD RESUME", /* couldn't resume download */

     "Error Curl Library: FILE COULDNT READ FILE",

     "Error Curl Library: LDAP CANNOT BIND",
     "Error Curl Library: LDAP SEARCH FAILED",
     "Error Curl Library: LIBRARY NOT FOUND",
     "Error Curl Library: FUNCTION NOT FOUND",
  
     "Error Curl Library: ABORTED BY CALLBACK",
     "Error Curl Library: BAD FUNCTION ARGUMENT",
     "Error Curl Library: BAD CALLING ORDER",

     "Error Curl Library: HTTP PORT FAILED", /* HTTP Interface operation failed */

     "Error Curl Library: BAD PASSWORD ENTERED", /* when the my getpass() returns fail */
     "Error Curl Library: TOO MANY REDIRECTS", /* catch endless re-direct loops */      
     "Error CURL LAST - Unknown"
   };
   
   if (error_code > CURLE_OK && error_code < CURL_LAST)
   {
      cout << endl << error_print_format[error_code + 1] << endl;
   }
}
/****************************************************************************/
bool download_file(const char *url_file, const char *tofile, either_file_or_directory filetype,
const char *proxy, const char *proxy_usernamepassword, bool use_tunnel)
{
  CURL *curl;
  CURLcode res;
  FILE *contentfile;
  bool return_value = false;
  
  contentfile = fopen(tofile, "w");
  curl = curl_easy_init();
  if (curl)
  {
    /* what call to write: */
    if (strcmp(proxy,"") != 0)
    {
       curl_easy_setopt(curl, CURLOPT_PROXY,proxy);
       if (strcmp(proxy_usernamepassword,"") != 0)
       	  curl_easy_setopt(curl, CURLOPT_PROXYUSERPWD,proxy_usernamepassword);
       if (use_tunnel)
          curl_easy_setopt(curl, CURLOPT_HTTPPROXYTUNNEL, TRUE);
    }
    curl_easy_setopt(curl, CURLOPT_URL, url_file);
    curl_easy_setopt(curl, CURLOPT_FILE, contentfile);
    curl_easy_setopt(curl, CURLOPT_NOPROGRESS, TRUE);
    if (filetype == is_a_directory)
    {
       curl_easy_setopt(curl, CURLOPT_FTPLISTONLY, TRUE);
    }
    res = curl_easy_perform(curl);
    /* always cleanup */
    curl_easy_cleanup(curl);
    print_curl_error(res);
    return_value = (res==CURLE_OK);
  }
  fclose (contentfile);
  return return_value;
}

